File: packages/android-compose/src/main/kotlin/org/digitalgreen/farmerchat/compose/network/TokenStore.kt
package org.digitalgreen.farmerchat.compose.network

/**
 * In-memory store for JWT tokens and device identity.
 * All state is held only for the lifetime of the process — the SDK has no local persistence.
 *
 * Thread-safe via @Volatile + synchronized writes.
 */
internal object TokenStore {

    @Volatile var accessToken: String = ""
        private set

    @Volatile var refreshToken: String = ""
        private set

    @Volatile var userId: String = ""
        private set

    @Volatile var deviceId: String = ""
        private set

    /** ISO country code returned by initialize_user (e.g. "IN"). Used for language queries. */
    @Volatile var countryCode: String = ""
        private set

    /** Country name returned by initialize_user (e.g. "India"). */
    @Volatile var country: String = ""
        private set

    /** State name returned by initialize_user (e.g. "Karnataka"). */
    @Volatile var state: String = ""
        private set

    @Volatile var isInitialized: Boolean = false
        private set

    @Synchronized
    fun saveTokens(
        accessToken: String,
        refreshToken: String,
        userId: String,
        countryCode: String? = null,
        country: String? = null,
        state: String? = null,
    ) {
        this.accessToken = accessToken
        this.refreshToken = refreshToken
        this.userId = userId
        if (!countryCode.isNullOrEmpty()) this.countryCode = countryCode
        if (!country.isNullOrEmpty()) this.country = country
        if (!state.isNullOrEmpty()) this.state = state
        this.isInitialized = true
    }

    @Synchronized
    fun saveAccessToken(accessToken: String, refreshToken: String) {
        this.accessToken = accessToken
        if (refreshToken.isNotEmpty()) {
            this.refreshToken = refreshToken
        }
    }

    @Synchronized
    fun setDeviceId(id: String) {
        this.deviceId = id
    }

    @Synchronized
    fun clear() {
        accessToken = ""
        refreshToken = ""
        userId = ""
        countryCode = ""
        country = ""
        state = ""
        isInitialized = false
        // deviceId is intentionally preserved across clear() — it must never change
    }
}

File: packages/ios-uikit/Sources/FarmerChatUIKit/Network/TokenStore.swift
import Foundation

/// In-memory store for JWT tokens and device identity.
/// Thread-safe via Swift `actor`.
actor TokenStore {

    static let shared = TokenStore()
    private init() {}

    private(set) var accessToken:  String = ""
    private(set) var refreshToken: String = ""
    private(set) var userId:       String = ""
    private(set) var deviceId:     String = ""
    private(set) var isInitialized: Bool  = false

    // IP-geolocation from initialize_user (no GPS permission needed)
    private(set) var countryCode: String = ""
    private(set) var country:     String = ""
    private(set) var state:       String = ""

    func saveTokens(
        accessToken:  String,
        refreshToken: String,
        userId:       String,
        countryCode:  String? = nil,
        country:      String? = nil,
        state:        String? = nil
    ) {
        self.accessToken  = accessToken
        self.refreshToken = refreshToken
        self.userId       = userId
        self.isInitialized = true
        if let cc = countryCode, !cc.isEmpty { self.countryCode = cc }
        if let c  = country,    !c.isEmpty   { self.country = c }
        if let s  = state,      !s.isEmpty   { self.state = s }
    }

    func saveAccessToken(_ access: String, refreshToken: String) {
        self.accessToken = access
        if !refreshToken.isEmpty { self.refreshToken = refreshToken }
    }

    func setDeviceId(_ id: String) { self.deviceId = id }

    func clearTokens() {
        accessToken  = ""
        refreshToken = ""
        userId       = ""
        isInitialized = false
    }
}
